CMakeModules/zeroconf.cmake

```diff
@@ -0,0 +1,27 @@
+find_package(DNSSD)
+find_package(Threads REQUIRED)
+
+find_library(AVAHI_COMMON_LIBRARIES avahi-common)
+find_library(AVAHI_CLIENT_LIBRARIES avahi-client)
+
+if (AVAHI_COMMON_LIBRARIES)
+	add_library(libavahi ${PROJECT_SOURCE_DIR}/third_party/avahi-compat-libdns_sd/compat.c)
+	target_link_libraries(libavahi PUBLIC ${AVAHI_CLIENT_LIBRARIES} ${AVAHI_COMMON_LIBRARIES})
+	target_compile_definitions(libavahi PUBLIC -DDNSSD_AVAHI)
+	set(HAS_AVAHI_SUPPORT ON)
+endif()
+
+
+if (DNSSD_FOUND)
+	set(HAS_ZEROCONF_SUPPORT ON)
+	add_library(zeroconf INTERFACE)
```

Comment:
Why is this library not prefixed by aseba?

Reply:
It's not related to aseba.
aka it's not asebazeroconf - which also exists.

when linking to `zeroconf` a target will automatically get the includes and libraries defined by `zeroconf` aka DNSSD_LIBRARIES, AVAHI_LIBRARIES, thread, support, etc.

it exists only to aggregate properties under a single name.

Reply:
Ok, thanks for the explanation. Another question: couldn't these properties directly be attached to asebazeroconf?

Reply:
We could but I think it's cleaner that way, more modular.
Asebazeroconf depends on zeroconf and zeroconf happens to be a set of dnssd / avahi library.
It helps hiding the nastiness of find packages so that we can reason in terms of libraries where a library is a set of includes, defines, properties, compiler flags and actual library ( .a / .so ).

A library can either be private or public, which I haven't worked a lot on yet. That controls whether the properties get inherited transitively or not.

Reply:
Ok, that seems to make sense. Do you plan to document these design decisions?

Reply:
I can put a command in the CMakefile once its done, or maybe more comment in the issue ?
CMake has the ability to generate a graphviz .dot from the dependencies tree Maybe that's a good resource ?
You are probably more experienced with graphviz than me, I'm not sure how to make the graph reasonably good looking.

Reply:
Comments are always good, as is a dependency tree. But the key elements would be a description of the rationale behind the organization. Obviously you re-organize the tree with some concepts in mind, it would be helpful to explain them so that people can understand the CMake files easily. The current one are very unsophisticated, so it is not so much of a problem for people to understand them quickly, but with more sophisticated ones, a design documentation is necessary.

Reply:
Note that I'm merely trying to implement what I understand to be good "modern cmake" practices  namely
 - Describe everything as a dependency tree
 - Use as little compilers flags as possible, warning levels, optimization levels, sanitizers, etc should be handled in CTest files.

Reply:
This is already a start of a design documentation :-)

Reply:
Good choices are bubble graphs or chord diagrams, depending on the topology. DependencyWheel from D3 is a typical choice for chord diagrams.